doc: replace mz_logical_timestamp with mz_now (#15039)
MaterializeInc/materialize: doc/user/content/sql/functions/now_and_mz_now.md

> ---

Reviewer: This is my first time reading this document. I know this isn't your work, but it seems very off base. Left you a number of line edits, sorry!

> title: "now and mz_now functions"

Author: This file has content changes beyond search-and-replace. It's way better looking now without all the extract and _ms stuff!

> description: "In Materialize mz_now() acts the way now() does in most systems."

Reviewer: ```suggestion
description: "Details the differences between the `now()` and `mz_now()` functions."
```

menu:
  main:
    parent: 'sql-functions'
---

> In Materialize, `now()` doesn't represent the system time, as it does in most systems; it represents the time with timezone when the query was executed. It cannot be used when creating views.
> 
> `mz_now()` comes closer to what `now()` typically indicates. It represents the logical time at which a query executes. Its typical uses are:

Reviewer: ```suggestion
In Materialize, `now()` returns the value of the system clock with timezone when the transaction began. It cannot be used when creating views.

By contrast, `mz_now()` returns the logical time at which the query was executed. This may be arbitrarily ahead of or behind the system clock.

For example, at 9pm, Materialize may choose to execute a query as of logical time 8:30pm, perhaps because data for 8:30–9pm has not yet arrived. In this scenario, `now()` would return 9pm, while `mz_now()` would return 8:30pm.

Its typical uses are:
```

> * Internal debugging

Reviewer: ```suggestion
* Query time introspection

  `mz_now()` can be useful if you need to understand how up to date the data returned by a query is. The data returned by a query reflects the results as of the logical time returned by a call to `mz_now()` in that query.
  
* [Temporal filters]

  Various windowing idioms involve using `mz_now()` in a [temporal filter](/sql/patterns/temporal-filters).
```

* Defining [temporal filters](/sql/patterns/temporal-filters/)

> ## Internal debugging
> 
> `mz_now()` can be useful if you need to understand the time up to which data has been ingested by `materialized`. It corresponds to the timestamp column of [`SUBSCRIBE`](/sql/subscribe). Generally a `CREATE SOURCE` command will cause Materialize to ingest data and produce timestamps that correspond to milliseconds since the Unix epoch.
> 
> ## Temporal filters
> 
> You can use `mz_now()` to define temporal filters for materialized view, which implement various windowing idioms.
> 
> For more information, see [Temporal Filters](/sql/patterns/temporal-filters/).

Reviewer: ```suggestion
```

## Restrictions

You can only use `mz_now()` to establish a temporal filter in one of the following types of clauses:

* WHERE clauses, where `mz_now()` must be directly compared to [`mztimestamp`](/sql/types/mztimestamp) expressions not containing `mz_now()`

> * A conjunction (AND), where `mz_now()` must be directly compared to [`mztimestamp`](/sql/types/mztimestamp) expressions not containing `mz_now()`.

Reviewer: ```suggestion
You can only use `mz_now()` to establish a temporal filter within a `WHERE` clause in which `mz_now()` must be directly compared to [`mztimestamp`](/sql/types/mztimestamp) expressions not containing `mz_now()`. The comparison must be the only expression in the `WHERE` clause, or in a conjunction (AND) chain at the top level of the `WHERE` clause.

Reviewer: Wait, no, this entire section is duplicative with the temporal filters page.

>   At the moment, you can't use the `!=` operator with `mz_now` (we're working on it).

Reviewer: ```suggestion
At the moment, you can't use the `!=` operator with `mz_now`.
```

> ## Restrictions
> 
> You can only use `mz_now()` to establish a temporal filter in one of the following types of clauses:
> 
> * WHERE clauses, where `mz_now()` must be directly compared to [`mztimestamp`](/sql/types/mztimestamp) expressions not containing `mz_now()`
> * A conjunction (AND), where `mz_now()` must be directly compared to [`mztimestamp`](/sql/types/mztimestamp) expressions not containing `mz_now()`.
> 
>   At the moment, you can't use the `!=` operator with `mz_now` (we're working on it).

Reviewer: This is mostly duplicated with the temporal filters page, except it is missing some detail. Let's just delete it?

```suggestion
```

## Example

### Temporal filter using mz_now()

<!-- This example also appears in temporal-filters -->

> For this example, you'll need to create a sample data source and create a materialized view from it for later reference.

Reviewer: ```suggestion
For this example, you'll need to create a sample data source and create an indexed view from it for later reference.
```

```sql
--Create a table of timestamped events.
CREATE TABLE events (
    content text,
    insert_t timestamp,
    delete_t timestamp
);

> --Create a materialized view of events valid at a given logical time.
> CREATE MATERIALIZED VIEW valid AS
> SELECT content, insert_t, delete_t
> FROM events
> WHERE mz_now() >= insert_t
>   AND mz_now() < delete_t;
> ```

Reviewer: This should probably be a view with a default index, right?
```suggestion
--Create an indexed view of events valid at a given logical time.
CREATE VIEW valid AS
SELECT content, insert_t, delete_t
FROM events
WHERE mz_now() >= insert_t
  AND mz_now() < delete_t;
CREATE DEFAULT INDEX ON valid;
```

Next, you'll populate the table with timestamp data.

> The epoch extracted from `now()` is measured in seconds, so it's multiplied by 1000 to match the milliseconds in `mz_now()`.

Reviewer: ```suggestion
```

```sql
INSERT INTO events VALUES (
    'hello',
    now(),
    now() + '100s'
);
INSERT INTO events VALUES (
    'welcome',
    now(),
    now() + '150s'
);
INSERT INTO events VALUES (
    'goodbye',
    now(),
    now() + '200s'
);
```

> Then, before 100 seconds (or 1.67 minutes) elapse, run the following query to see all the records:

Reviewer: ```suggestion
Then, before 100 seconds elapse, run the following query to see all the records:
```

```sql
SELECT *, mz_now() FROM valid;
```
```nofmt
 content |        insert_t         |        delete_t         |    mz_now
---------+-------------------------+-------------------------+---------------
 hello   | 2022-09-27 23:52:34.831 | 2022-09-27 23:54:14.831 | 1664322794280
 goodbye | 2022-09-27 23:53:04.262 | 2022-09-27 23:56:24.262 | 1664322794280
 welcome | 2022-09-27 23:53:03.142 | 2022-09-27 23:55:33.142 | 1664322794280
(3 rows)
```

> If you run this query again after 1.67 minutes from the first insertion, you'll see only two results, because the first result no longer satisfies the predicate.

Reviewer: ```suggestion
If you run this query again after 100 seconds from the first insertion, you'll see only two results, because the first result no longer satisfies the predicate.
```

### Query using now()

> The epoch extracted from `now()` is measured in seconds, so it's multiplied by 1000 to match the milliseconds in `mz_now()`.

Reviewer: ```suggestion
```
